src/python/pants/engine/README.md

```diff
@@ -49,44 +50,44 @@ then the `@rule` will run exactly once, and the value that it produces will be a
 
 #### Example
 
-As a very simple example, you might register the following `@rule` that can compute a `String`
-Product given a single `Int` argument.
+As a very simple example, you might register the following `@rule` that can compute a `str`
+`Product` given a single `int` argument.
 
 ```python
-@rule(str, [int])
-def int_to_str(an_int):
-  return str(an_int)
+@rule
+def int_to_str(value: int) -> str:
+  return str(value)
 ```
 
-The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
-second argument is a list of "parameter selectors" that declare the types of the input parameters for
-the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
-selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
+Here the only requirements for the `@rule` are:
+1. The decorated function is a free, top-level function; ie: methods cannot be `@rule`s.
+2. The decorated function must be fully type annotated.
```

Comment:
I would like to keep this simple, unequivocal language, and it seems that if we do, we would also want to actually deprecate the old `@rule` syntax (without annotations) in this PR as well? Or are we referring to either syntax here?

Reply:
I think its fine that the language is simple without the deprecation in the same PR. In docs we steer folks to use annotations by not mentioning the soon to be deprecated syntax. No need to mention the old syntax since the new syntax covers everything needed by a user.

Reply:
Hm, ok. I might decide to avoid deprecating `datatype` in the same PR as #8181 then to unblock it!